libs/knowledge-store/ragstack_knowledge_store/document_graph.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+import itertools
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Optional, Union, Iterator, Tuple
+
+from langchain_core.callbacks import CallbackManagerForRetrieverRun
+from langchain_core.documents import Document
+from langchain_core.retrievers import BaseRetriever
+from pydantic import Field
+
+
+def _texts_to_documents(
+    texts: Iterator[str],
+    metadatas: Iterator[dict],
+) -> Iterator[Document]:
+    for (
+        text,
+        metadata,
+    ) in zip(texts, metadatas):
+        yield Document(
+            page_content=text,
+            metadata=metadata,
+        )
+
+
+class DocumentGraph(ABC):
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[Iterable[dict]] = None,
+        *,
+        ids: Optional[Iterable[str]] = None,
+        links: Iterable[Tuple[str, str]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        metadatas_it = iter(metadatas) if metadatas else itertools.repeat({})
+        documents = _texts_to_documents(iter(texts), metadatas_it)
+        return self.add_documents(documents, ids=ids, links=links, **kwargs)
+
+    @abstractmethod
+    def add_documents(
+        self,
+        documents: Iterable[Document] = None,
+        *,
+        ids: Optional[Iterable[str]] = None,
+        links: Iterable[Tuple[str, str]] = None,
```

Comment:
I don't know that these will be enough to create links. Can you elaborate on how this would work?

Reply:
The tuple is (from_id, to_id). This is like this for simplicity but maybe adding a `Link` class that can be extended would be better.

Reply:
There's a few difficulties with making that form useful.

1. IDs may not have been assigned
2. Different kind of link inferences may rely on things being written first / DB queries / etc. I'm not sure it would be possible to infer all of the links ahead of time (may need to compute embeddings, etc.).

I wonder if we could make this use #450 -- basically, we could have an edge extractor that looked for explicit links in the metadata. This would be a list of pairs like this. Then, the user could choose to have that edge extractor or not?

Reply:
If we're going to keep the `links` parameter, it's probably worth creating an `Edge` class to represent this tuple, just so it's clear what the schema is and we have a place to document things.

That being said, I think I would prefer to hide the edges and have them be created by the store based on a configured set of extractors (I think this is #450?)

Reply:
This proposal is compatible with #450 as edge extractors are configured in the implementation init and we pass the metadata in the `add_documents`/`add_texts` methods.
Ok to create `Link` or `Edge` rather than `Tuple[str, str]`.
The idea here is to support edges/relationships created outside of the DB. For instance, from a process that would build the full graph externally and then dump it to the DB. This would have the advantage of being DB independent.

Reply:
Edge extractors are very tied to Cassandra, so I don't think it's possible to pull them to the main API atm.

Reply:
Removed links from the API for the moment as explicit links can be defined in metadata (see ExplicitEdgeExtractor)